### ISSUE TITLE
Assortment of Ridge Racer bugs

### ISSUE THREAD

**Issue by ghost** [hrydgard/ppsspp#13194]
The new update for PPSSPP already came but some bugs in Ridge Racer 1 and 2 are left untouched and isn't even fixed or will never be fixed at all... 

Already discovered bugs:

- Broken speedometers (Bayonet, Esperanza, Fatalita, Angelus Mk1, Crinale Mk1, Crinale Mk2, Pac-Man Mk2) #8509
- AV Player desyncing (The pre-recorded A.I. Replays are bugged and car hits wall) #2990 
- Cutscenes suddenly stops abruptly (The cutscenes stops 1-6 seconds earlier) #11934 
- Nitrous visual glitch (Still Images of buildings from the race track appearing in a almost transparent state) #11915 
- Multiplayer won't work (The workaround for the multiplayer does not work for me) #10698 

Newly found bugs:

- Headlights glow effect (When car lights are on, they have a white star shaped effect on normal PSP but not on PPSSPP)
- Black box on menu (Black boxes appear sometimes when scrolling on the menu)
- Black lines on menu (Black lines appear on menus, disappears when I set Rendering Resolution to 1x PSP)

Honestly, these bugs will never be fixed and will be forever ignored...

**Comment by hrydgard** [comment]
I don't know where you're getting the "will never be fixed" thing from. In an emulator, generally we fix systems, not games. It might be that a future rewrite of the render target system will fix several of these, there's no "never".

As for the replays, those have already been reported in a separate issue and it's an insanely tricky problem.

I'll leave this open since it contains some "newly found bugs" and not just repeats.

As for the headlights glow, got a screenshot of how it looks and what it should look like? Not familiar with that one (normally you mostly see cars from behind anyway...).

Black boxes I haven't seen in the menus. On what device are you seeing that? Which GPU?

**Comment by GrandGTA** [comment]
I am using an NViDiA Quadro K620, a decent Intel CPU, and 32GB of RAM... The fact that I said _"will never be fixed"_ is because that months ago, when I first started reporting RR7 bugs in the RPCS3 Github, one RPCS3 developer is fully toxic to me... I reported the same AV Player bug which happens in Ridge Racer 7 and I gave all my info about the bug but he just Closed the issue and disliked it... Heck, he even added a tag called "Meta: Invalid" to my issues, he didn't even asked _"I don't fully understand the bug your experiencing.. please specify it a little bit more.."_ Instead, he just closed my issues as If he thinks its just garbage... I asked him when will the Sunset Heights Red Bug get fixed (already fixed) but he responded with _"where can I get that sh*t you're smoking"_ as if he is saying I'm mentally ill or just on drugs... I got really triggered and stayed off the RPCS3 Github... Hey man, at least your kind and not toxic like him and you always keep an issue open until it is fixed... If your wondering who was going toxic on me, it was this craphead called **"AniLeo"**...

**Comment by boy13epic** [comment]
Thats a pretty good list of bugs

**Comment by HiImBrala** [comment]
@GrandGTA
What i'm going to say is actually my opinion,and i hope you will understand.Issue here is not guy making emulator,but his surrounding itself,because people act as they see,something like reverse engineering at this point.If they see someone acting like a *craphead*,they will do it because they like their behaviour.That is issue of him being "*pressured*",and acting like that will just make him bad reputated,what actually can be his goal at first.People don't realize when they make something that involves people,they shouldn't act like that and they should *"grow up"*,because that is priority and rensposibility in  work like that.In fact,always has been like that.I won't name people i don't think are here,but act as BS towards others because you have problems won't really solve anything,and it can be trigger to others.Mental health issues are not joke,him saying something like that is really awful.If you ask me,i would spread this among anyone who wants to do same in the future.

**In short:screw morons that don't have sense of rensposibility and kindness,as i reported and was reporting at few developers,and i haven't met anyone that rude with mental complex like that idiot,and everyone deserves to know what BS he pulls because of that.Cheers**

**Comment by hrydgard** [comment]
@GrandGTA RPCS3 is a completely different project with different people.

**Comment by GrandGTA** [comment]
@ghost Thanks mate... At least you care about cyberbullying.... :)

@hrydgard Yes, I know that RPCS3 is different but they have same issues in Ridge Racer like AV Player Bugs... The difference about the community and people is they just ignore those bugs unlike the community here in PPSSPP, the PPSSPP community here is just calm and kind and is willing to fix bugs... Speaking of bugs, check out the Greenpeak Highlands replay...

The star shaped effect in the headlight is much more noticable:

![Headlight Stars](https://user-images.githubusercontent.com/56748157/88455142-897fc800-cea5-11ea-94bd-3c2ed5e9e08a.png)

And I also noticed that the screen is much more stretched in the normal PSP hardware:

Original PSP Hardware:
![Original PPSSPP Hardware](https://user-images.githubusercontent.com/56748157/88455215-1b87d080-cea6-11ea-8708-4a9622d3e34c.png)

PPSSPP Hardware:
![PPSSPP Hardware](https://user-images.githubusercontent.com/56748157/88455231-378b7200-cea6-11ea-8b7b-9720458d809f.png)

If you're gonna ask me if it probably was the recording device because the picture is stretched but no... When I first tried PPSSPP after my PSP got broken, the first thing I noticed is that I am not comfortable with the screen because my PSP displays the game screen stretched... I checked out a gameplay of Ridge Racer on a normal PSP and I was right... The screen is normally stretched...

The lines im talking about is much more visible in the New Rally-X Minigame:
![Mini-Game](https://user-images.githubusercontent.com/56748157/88455363-50e0ee00-cea7-11ea-9fd1-c4acb29ac2cb.PNG)

The black boxes, im not experiencing them again... It was something i did with the settings and it just disappeared...

**Comment by LunaMoo** [comment]
Artifacts related to rendering above x1 res are normal with most 2D surfaces as PSP games were not designed nor tested for running above PSP res. You can sometimes counter it with texture scaling or better texture replacement, but there are opposite cases where at least automatic texture scaling will only make it worse. You can also counter it by forcing nearest texture filtering, but it will make things look bad and off from real thing. Either way it's not something that can be "fixed", so if workarounds doesn't help/aren't acceptable, you will have to run the game with x1 render res or live with the glitches, through it's possible to deal with it differently ie using custom upscaling filter instead of linear filtering.

**Comment by GrandGTA** [comment]
I noticed that both Android and PC loads the AV Player differently causing it to desync differently... List out the differences on how both of these platforms load and execute the AV Player CPU Autodrive and you might find the fix for it...

**Comment by GrandGTA** [comment]
Right now, some of the people in the RPCS3 Github got influenced with **AniLeo's** idiotic behavior towards me calling me a retard or a piece of shit.... Honestly, i'm really goddamn pissed at seeing this.... **AniLeo** is trying to make me look like a bitch and is influencing people that he is a good person and I am the piece of skunk shit.... I'm going to report him rn...

**Comment by HiImBrala** [comment]
> Right now, some of the people in the RPCS3 Github got influenced with **AniLeo's** idiotic behavior towards me calling me a retard or a piece of shit.... Honestly, i'm really goddamn pissed at seeing this.... **AniLeo** is trying to make me look like a bitch and is influencing people that he is a good person and I am the piece of skunk shit... He also called me "expired milk".... I'm going to report him rn...

As i said,guy is moron,and his presence is poisonus to every human being there,no point in fighting against them,when it's clear you're better than them.

These bugs persist on my PC,really good list.I can do investigation to find more.

**Comment by CookiePLMonster** [comment]
I'll just say one thing and not respond further to avoid cluttering the topic and causing headaches for the PPSSPP team, but IMHO the fact you show up with a "reporting issues which will never be fixed lol" attitude and then immediately proceed to complain about being mistreated elsewhere **does not** paint a good first impression. If your issue in the RPCS3 repo was kept in a similar "fix this already!" tone as this one is, I am not surprised in the slightest it was met with a negative response.

**Comment by GrandGTA** [comment]
Im gonna close this issue... These bugs aren't a big deal anyways since it does not interfere with the gameplay itself...

**Comment by GrandGTA** [comment]
I deem the AV Player SR765 Replay impossible for me to replicate... Used a CE speedhack, the Angelus always throws me off by overtaking or not taking a specific lane...

**Comment by gamer-gamer** [comment]
Maybe an option not to upscale 2d like Gliden64 for n64 emulator does might correct some of those 2d issues ? Unless that isnt possible.

**Comment by GrandGTA** [comment]
Also found out that since Android and PC have different ways of desyncing because of different CPU numbers and calculations, try and list every and I mean every single CPU differences on both Android and PC version of PPSSPP... Its like Android is 1.0 and PC is 2.0 and to stop it from desyncing, it needs to be in 1.5...

**Comment by Back2Life888** [comment]
Hey guys, I got a better image of this "Headlight Star thing"

![Sparkol](https://user-images.githubusercontent.com/76780043/105632784-0fe68800-5e90-11eb-9ae2-e8ee264d8a05.PNG)

**Comment by Back2Life888** [comment]
Hey guys, I have a better image than GrandGTA. I found a Ridge Racer PSP video with a good recording device and resolution of 720p.. This is what it looks like with much better quality.

![Sparkol](https://user-images.githubusercontent.com/76780043/105632813-2b519300-5e90-11eb-920c-1db2a8b9ea03.PNG)

**Comment by Back2Life888** [comment]
Im also planning on getting a new PSP to contribute to these bugs..

**Comment by Veritas83** [comment]
I think i found a workaround... I just need to confirm it now.

https://i.gyazo.com/f3f36fa3aea13a49d93b8ce6dc453c6c.png

**Comment by hrydgard** [comment]
That is interesting! Might help figuring out what's going wrong here.

**Comment by Veritas83** [comment]
I've also found they headlights will shine anytime they are supposed to thru the regular viewport (main game cam) with 1x resolution.

Not sure if they are supposed to appear in rearview mirror. But if i drift particularly enthusiastically thru a tunnel. The opponents headlights are shining

**Comment by Veritas83** [comment]
Possibly related.

I've been testing [ReinterpretFramebuffers] & [ShaderColorBitmask] with this title UCES00422

And thus far I've noticed no adverse effects.

**Comment by hrydgard** [comment]
Have you noted any positive effects of those, though?

**Comment by Veritas83** [comment]
No, Reflections may be improved. I'll have to compare some streams side by side (twitch.tv/Veritas0923)

I also noticed one more thing. A white-ish box around the tach with Assoluto

<img="https://i.gyazo.com/d469aff712ae3917e8a25c18ec6f3d4f.png">

I've been trying all manner of settings to fix headlights, digital tachs and that white box specifically.

**Comment by Back2Life888** [comment]
> I think i found a workaround... I just need to confirm it now.
> 
> Ok. This is confirmed now.
> 
> Setting 'Internal Resolution' to anything other than 1x will break the headlights.
> 
> 1x Internal Resolution. Texture Scaling: 0
> 
> ![](https://camo.githubusercontent.com/868be835242716c46c81dd0b5a0a77c6084d4388c353cb00004be06eefdfd2f8/68747470733a2f2f692e6779617a6f2e636f6d2f66336633366661336165613133613439643933623863653664633435336336632e706e67)
> 
> 1x Internal Resolution Texture Scaling: 3 Video Filter: hq3x
> 
> ![](https://camo.githubusercontent.com/69679e7d3b80fad58fdc7737a911d961d0c7d8fd89f4328d8c844c9fbd7d449f/68747470733a2f2f692e6779617a6f2e636f6d2f65343334616134643936376530613864303332663632363435613133613062662e706e67)

hold on.. How in the hell did you manage to load up the headlight glare effect even if PPSSPP doesn't load it??

**Comment by Veritas83** [comment]
By using the settings specified. Any setting for 'Internal Resolution' other than 1x (the first one) will break the headlight glare. Scale by other means (hq2x/hq3x video shader or example) the headlight glare remains.

**Comment by Back2Life888** [comment]
Where can I find the hq2x or hq3x video shader?? And by Internal Resolution, you mean Rendering Resolution??

**Comment by Veritas83** [comment]
In RetroArch under PPSSPP Specific Core Options it is called 'Internal Resolution' 

In the standalone PPSSPP it is effected by both Window Size and Render Resolution. Make sure Render Resolution is 1x

**Comment by Back2Life888** [comment]
Im using a standalone PPSSPP and not the RetroArch PPSSPP.. So I don't really understand you since you speak RetroArch PPSSPP language and not standalone PPSSPP language.. Try recreating the workaround for the headlight glare with the standalone PPSSPP.. I think RetroArch behaves more differently than the Standalone PPSSPP for PC..

**Comment by Veritas83** [comment]
Just got it working in standalone as well.

<img src="https://i.gyazo.com/92f22433ada9503d49a533b17b1522bd.png">

On top of 1x Resolution, It also requires OpenGL Driver mode. D3D11 didn't work for me.

**Comment by Back2Life888** [comment]
OpenGL?? So thats f***ing why.. You didn't mention about OpenGL..

**Comment by Veritas83** [comment]
I didn't know it required OpenGL until just now. RetroArch defaults to OpenGL

**Comment by Back2Life888** [comment]
Nevermind.. It still never worked.. How about you take screenshots of your settings??

**Comment by Veritas83** [comment]
No headlights in RA_PPSSPP w\ d3d9 or d3d11

Here's my standalone settings.

<img src="https://i.gyazo.com/bd95184e67a25628d27513dabc89a8f4.png">

<img src="https://i.gyazo.com/801106f3c00cb2536e48caa9cc09dbc5.png">

**Comment by Back2Life888** [comment]
I think it was my PPSSPP's version that caused it.. Mine is 1.10.3 but yours is 1.11.2..

**Comment by Veritas83** [comment]
Oh well, jeez man. Always run the latest version. That's how you get the latest fixes....

**Comment by Back2Life888** [comment]
Still not working.. I guess you're lucky..

**Comment by Veritas83** [comment]
I think the only other thing I ticked was 'Vertex Shading' because that was enabled by default in RetroArch as well. Otherwise I'm using default settings. Never even used standalone.

Maybe try RetroArch?

**Comment by Veritas83** [comment]
Reflections do seem improved in my RetroArch over what little I've seen in standalone (no compat.ini options set there)

**Comment by Back2Life888** [comment]
![Good](https://user-images.githubusercontent.com/76780043/108053751-ca7d3b00-7088-11eb-881f-248e82054f88.PNG)
Weird.. It works only on RetroArch.. Henrik needs to check whats going on..

**Comment by Back2Life888** [comment]
I need to know where did you download your PPSSPP version.. I think you got a very unique PPSSPP build..

**Comment by Veritas83** [comment]
Just directly from the homepage. Same thing anyone else would be using that is using standalone (and latest version)

**Comment by Back2Life888** [comment]
One last question.. Is RetroArch in the same folder in PPSSPP??

**Comment by Veritas83** [comment]
lol no, C:\EMU\RetroArch\ & C:\EMU\PSP\

**Comment by Back2Life888** [comment]
Well in conclusion.. Your specs may be s**t af while my specs are better than yours but somehow, your specs brings back one of the most nostalgic effect on the game..

**Comment by Back2Life888** [comment]
I guess I give up on this one..

**Comment by Veritas83** [comment]
It's not the size of your GHz that counts ... It's how you use it...

**Comment by Back2Life888** [comment]
but still.. You managed to make it work.. I give up since it probably wouldn't work..

**Comment by Veritas83** [comment]
Enthusiastically drifting thru a tunnel during a race... Headlights observed.

<img src="https://i.gyazo.com/77cf372d756dc8c6a7e26385336db986.png">

**Comment by Veritas83** [comment]
ReinterpretFramebuffers & ShaderColorBitmask On

<img src="https://i.gyazo.com/2613c502a73080a2a1377099f4e05635.png">

ReinterpretFramebuffers & ShaderColorBitmask Off

<img src="https://i.gyazo.com/e3c27cbe6f82d221c15ef6138d796ea9.png">

**Comment by hrydgard** [comment]
However the game is doing the headlights, it's doing something very delicate or sensitive, it seems, since it breaks so easily. Can be tiny differences between how different GPUs render. Need to figure out how it's actually working to fix it properly...

**Comment by Veritas83** [comment]
The 1x resolution & OpenGL seem key to making it work thru the regular game race/countdown/replay. Trying to make it work in rearview now...

Basically trying every setting in compat.ini w\ 1x res now.

**Comment by Veritas83** [comment]
I think this, may be a relevant frame. Stepping thru the textures this is the first time the headlights appear in rearview.

<img src="https://i.gyazo.com/d38a3a052811e120e81d97c3c0d1ad7b.png">

This is the rearview itself being drawn

<img src="https://i.gyazo.com/829f1dfd817f8ae7bffbc524cc286443.png">

**Comment by Veritas83** [comment]
Screen being copied

<img src="https://i.gyazo.com/562c386f1b5878287642f0c6da9354df.png">

HUD elements about to be drawn

<img src="https://i.gyazo.com/3689ca230aaf128987035f2adad8e4ac.png">

**Comment by Veritas83** [comment]
Here's a savestate of me paused at that same frame.

[PPSSPP_STATE.zip](https://github.com/hrydgard/ppsspp/files/5989725/PPSSPP_STATE.zip)

**Comment by Veritas83** [comment]
Before i was stepping by frame. Stepping into each instruction while it builds the frame ... This is where it draws the actual headlight

<img src ="https://gyazo.com/74b676ab0a3cfe104ceefa9a5afd5b9d.png">

<img src ="https://gyazo.com/2090f371d115251e6095a6efa677462d.png">

That part is applied before. Then whatever this is

<img src ="https://gyazo.com/e0f85d529792308862a1dd3ede164557.png">

Then a similar process repeats for the actual headlight

<img src ="https://gyazo.com/2e9f0f08517dce5ceb4103f0e2766e5d.png">

<img src ="https://gyazo.com/9dc4a57dd0fbf8d78bcc443ae2c32709.png">

**Comment by hrydgard** [comment]
Hm, it's somehow texturing from a 1x1 texture (which happens to share address with a bigger one, which is why what you see there is not 1x1). Maybe that's supposed to be a certain value from the framebuffer at the pixel location of the headlight, so that if it gets overdrawn with something, it won't have that value and the headlight won't be drawn. or something...

**Comment by unknownbrackets** [comment]
This games does other rendering to a palette, might be related to this 1x1 business.  Maybe we don't clamp to 1x1 properly, esp with aniso or etc.?  Though we should clamp with offset and size settings, we've mostly focused on clamping when size is larger than actual FB size...

-[Unknown]

**Comment by Back2Life888** [comment]
hrydgard.. You should check the headlight glare thoroughly because it only works on RetroArch and not on your standalone version of PPSSPP.. I think its something with the specs..

**Comment by Veritas83** [comment]
Back2Life do you use a nVidia GPU by chance ?

As I've shown, It worked for me just fine in standalone PPSSPP.

But. I was using 1x Res, OpenGL and i use an AMD GPU. It isn't all *that* rare to have rendering differences/issues between nVidia/AMD. I recall a similiar rendering issue on PS3 emulation with RR7, nVidia couldn't do trees right for the life of it. But it 'just worked' on AMD. 

I've always found AMD wins these battles thou, So that's why I use an AMD GPU

**Comment by hrydgard** [comment]
Not much point in speculation, this just needs to be investigated properly.

**Comment by Veritas83** [comment]
Feel free to send me (or just branch) a test build if you would like me to test anything with my setup :)

**Comment by Back2Life888** [comment]
Yeah I am using an NVIDIA GPU..

**Comment by Veritas83** [comment]
Happened to be playing the original RR for PSP and can confirm the Headlights 1x Internal Res + OpenGL workaround also works there.

<img src="https://i.gyazo.com/3d7699db8928354fe65dc429ca05160a.png">

**Comment by Back2Life888** [comment]
The headlight glare is supposed to be opaque and not translucent.. And also, the multiplayer for this game is finally fixed..

**Comment by Back2Life888** [comment]
any updates?

**Comment by Back2Life888** [comment]
Henrik, how about we extract the replay's AI from the game, translate and apply the inputs on a certain tool, and replicate the replay itself? Its kind of like a Tool Assist where you put some input of buttons that the tool will press or hold for a certain time.

**Comment by combo-pizza** [comment]
@unknownbrackets  Do you know about the textures getting corrupted in the software render? The tracks look normal then after maybe a lap or two the texture colors start to get weird. I've heard the game does some weird memory stuff before starting a track though which I'm guessing the HLE render doesn't even touch. Also Cosmo canyon in ridge racer 2 is causing a crash that takes out the entire emulator in the WIP x64 windows build.

**Comment by Back2Life888** [comment]
> @unknownbrackets Do you know about the textures getting corrupted in the software render? The tracks look normal then after maybe a lap or two the texture colors start to get weird. I've heard the game does some weird memory stuff before starting a track though which I'm guessing the HLE render doesn't even touch. Also Cosmo canyon in ridge racer 2 is causing a crash that takes out the entire emulator in the WIP x64 windows build.

Probably a corrupted game copy.

**Comment by combo-pizza** [comment]
Have you tried it with the software render? It should happen for you too. The games are also dumped and verified from my own UMDs which play fine on my real psp 1000 so I don't see how they could be corrupted.

**Comment by Back2Life888** [comment]
So in summary (04-01-22)

-Still has broken replays (If anyone has a normal PSP hardware, record the undesynced replays and send me the video. The links given above are from a terminated account.)
-Speedometers without needles are broken (Workaround is software rendering but its super slow)
-Semi-broken headlight glare (Headlight glare works on NViDiA now! Only problem is though is that it renders at a small distance and only right headlights works. Headlight glare still works like a charm in NViDiA.)
-Screen ghosting effect when using NOS (sometimes occur)

There is progress, but it might take months to fix these. Maybe even years.

**Comment by hrydgard** [comment]
I believe these are all fixed:

* Speedometers
* Lens flare / headlight glare / whatever you call it
* Ghosting bug when using Nitro

Remaining are replays and FMV issues, plus general issues with texture tiling in the menus, which is not specific to Ridge Racer, just something that happens when you render a game at a higher resolution than what it was designed for.

We have separate issues for the remaining problems, not setting much point in this meta-issue anymore so I'll close it.
